packages/server/lib/browsers/bidi_automation.ts

```diff
@@ -0,0 +1,270 @@
+
+export type NetworkBeforeRequestSentParametersModified = NetworkBeforeRequestSentParameters & {
+  request: {
+    destination: RequestDestination
+    initiatorType: RequestInitiatorType
+  }
+}
+
+// maps the network initiator to a ResourceType (which is initially based on CDP)
+const normalizeResourceType = (type: RequestInitiatorType): ResourceType => {
+  switch (type) {
+    case 'css':
+      return 'stylesheet'
+    case 'xmlhttprequest':
+      return 'xhr'
+    case 'img':
+      return 'image'
+    case 'iframe':
+      return 'document'
+      // for types we cannot determine, we can set to other.
+    case 'audio':
+    case 'beacon':
+    case 'body':
+    case 'early-hints':
+    case 'embed':
+    case 'frame':
+    case 'input':
+    case 'link':
+    case 'object':
+    case 'track':
+    case 'video':
+    case null:
+      return 'other'
+    default:
+      return type
+  }
+}
+
+export class BidiAutomation {
+  // events needed to subscribe to in order for our BiDi automation to work properly
+  static BIDI_EVENTS = [
+    'network.beforeRequestSent',
+    'network.responseStarted',
+    'network.responseCompleted',
+    'network.fetchError',
+    'browsingContext.contextCreated',
+    'browsingContext.contextDestroyed',
+  ]
+  private webDriverClient: WebDriverClient
+  private automation: Automation
+  private autContextId: string | undefined = undefined
+  // set in firefox-utils when creating the webdriver session initially and in the 'reset:browser:tabs:for:next:spec' automation hook for subsequent tests when the top level context is recreated
+  private topLevelContextId: string | undefined = undefined
+  private interceptId: string | undefined = undefined
+
+  private constructor (webDriverClient: WebDriverClient, automation: Automation) {
+    this.automation = automation
+    this.webDriverClient = webDriverClient
+
+    // bind Bidi Events to update the standard automation client
+
+    // Error here is expected until webdriver adds initiatorType and destination to the request object
+    // @ts-expect-error
+    this.webDriverClient.on('network.beforeRequestSent', this.onBeforeRequestSent.bind(this))
+    this.webDriverClient.on('network.responseStarted', this.onResponseStarted.bind(this))
+    this.webDriverClient.on('network.responseCompleted', this.onResponseComplete.bind(this))
+    this.webDriverClient.on('network.fetchError', this.onFetchError.bind(this))
+    this.webDriverClient.on('browsingContext.contextCreated', this.onBrowsingContextCreated.bind(this))
+    this.webDriverClient.on('browsingContext.contextDestroyed', this.onBrowsingContextDestroyed.bind(this))
+  }
+
+  setTopLevelContextId (contextId?: string) {
+    debug(`setting top level context ID to: ${contextId}`)
+    this.topLevelContextId = contextId
+  }
+
+  private async onBrowsingContextCreated (params: BrowsingContextInfo) {
```

Comment:
the bulk of what happens here is what happens similar to CDP
* making sure the prerequests send the correct information to the middleware, along with the AUT header
* Makes sure the AUT frame is associated correctly when contexts are created/destroyed